lago/subnet_lease.py

```diff
@@ -17,222 +17,505 @@
-
-    Args:
-        index (int): Third element of a dotted ip representation of the subnet,
-            for example, for 1.2.3.4 it would be 3
-
-    Returns:
-        None
-    """
-
-    lease_dir = config['lease_dir']
-    lease_file = os.path.join(lease_dir, '%d.lease' % index)
-    os.unlink(lease_file)
-
-
-def release(subnet):
-    """
-    Free the lease of the given subnet
-
-    Args:
-        subnet (str): dotted ip or network to free the lease of
-
-    Returns:
-        None
-    """
-    _release(int(subnet.split('.')[2]))
+    def __init__(self, store_path, subnet):
+        self._store_path = store_path
+        self._subnet = subnet
+        self._path = None
+        self._realise_lease_path()
+
+    @staticmethod
+    def from_lease_filename(store_path, store_template, lease_filename):
+        third_octet = lease_filename.split('.')[0]
+        return Lease(store_path, store_template.format(third_octet))
+
+    def _realise_lease_path(self):
+        ip = self.subnet.split('/')[0]
+        idx = ip.split('.')[2]
+        self._path = os.path.join(self._store_path, '{}.lease'.format(idx))
+
+    def to_ip_network(self):
+        return IPNetwork(self.subnet)
+
+    @property
+    def valid(self):
+        if self.exist:
+            return self.has_env
+        else:
+            return False
+
+    @property
+    def metadata(self):
+        with open(self.path) as f:
+            uuid_path, uuid = json.load(f)
+
+        return uuid_path, uuid
+
+    @property
+    def has_env(self):
+        return self._has_env()
+
+    def _has_env(self, uuid_path=None, uuid=None):
+        if not (uuid_path and uuid):
+            uuid_path, uuid = self.metadata
+
+        if not os.path.isfile(uuid_path):
+            return False
+
+        with open(uuid_path, mode='rt') as f:
+            if f.read() == uuid:
+                return uuid_path
+            else:
+                return False
+
+    @property
+    def exist(self):
+        return os.path.isfile(self.path)
+
+    @property
+    def path(self):
+        return self._path
+
+    @path.setter
+    def path(self, data):
+        self._path = data
+
+    @property
+    def subnet(self):
+        return self._subnet
+
+    @subnet.setter
+    def subnet(self, data):
+        self._subnet = data
+
+    def __str__(self):
+        return self.subnet
+
+
+class LagoSubnetLeaseException(utils.LagoException):
+    pass
+
+
+class LagoSubnetLeaseLockException(LagoSubnetLeaseException):
+    def __init__(self, store_path):
+        super(LagoSubnetLeaseLockException, self).__init__(
+            dedent(
+                """
+                Failed to acquire a lock for store {}.
+                This failure can be caused by several reasons:
+                1. Another 'lago' environment is using the store.
+                2. A stale lock was left in the store.
+                3. The group which owns the path to the store isn't 'lago',
```

Comment:
Maybe just say the user doesn't have access to the store? because if he installed from pip - he didn't necessary setup a lago group at all(and also we don't really know if that is the reason - just that we can't access it).